internal/services/frontend/decode.go

```
@@ -0,0 +1,67 @@
+
+// Input and output for a success case
+var views = "Defined, Observed"
+var response []string
+var expected []bool = []bool {true, false, false, true}
+
+// Input and output for a faiure case
+var badView = "AbaddefinedobservedValue" 
+var badexpected []bool = nil
+
+func decode(keywords []string, source string) ([]bool, error){
+	// First subtask is to get from source to test (convert the source string into the test string slice)
+	// hint: strings package probably has something like split and trim.
+
+	var test []string = strings.Split(source, ",")
+	
+	
+	// Third subtask is to loop through all the test slice elements.
+	outputkeywords := make([]bool, len(keywords))
+	// I need to create putput array of right size with all elements set to False
+	
+	for _, t := range test{
+		t2 := strings.TrimSpace(t)
+		for i, k := range keywords{
+		
+	// Second subtask is to figure out how to test one element in the test slice
+	// -- how to tell if test[0] matches any of the keywords elements, and which one.
+			if k == t2 {
+			//Set the output element of [i] to true 
+				outputkeywords[i] = true
+			
+				//	response = append(outputkeywords[i])
+
+			//else 
+
+			//return error
+			
+			}
+		}
+	}
+	for _, t3 := range test{
+		t4 := strings.TrimSpace(t3)
+		for j, h := range keywords{
+
+	// Output array, nil 
+	//How and where we catch the error of mismatch(Write code to just spot the error) write a loop return  an error (true or false)
+
+	// keywords and source are passed into decode and are immutable.
+	// 1st subtask creates this: test is created from the data in source
+	// 2nd subtack creates this: matches []bool that is created inside decode and 
+	// 2nd subtask starts this: ...filled in by the match tests, and then returned
+	// 3rd subtask completes this: ...filled in by the match tests, and then returned
+			if h != t4{
+				outputkeywords[j] = false
+			}
+			return badexpected, error ()
```

Did this work when you ran the two unit tests below?  I assume that TestDecode would have failed, but wanted to check.

This looks like it will always return a failure -- there's nothing to prevent line 63 from executing.  And the outputkeywords array should only be part of the success test - it doesn't need to be modified in this failure test.

The thing you're trying to verify is that if an element in the test array does not match against any element in the keywords array, you should return an error.  The thing is, though, that you can only test one element at a time, so you have to reverse the thinking - namely that if you find an element in the keywords array that matches the element from the test array then you can't fail on that test element.

So, think about putting a new declaration between lines 49 & 50 -- specifically, 'found := false'.  Now think about when found should be set to true, and when you should check what the state of found is in order to decide if you need to return an error.

---

[Also, don't return badexpected.  Return nil.  The declarations from lines 9 to 16 were to document what the two unit tests should be looking for]